### PR TITLE
Support Inductor collectives with wait or collective outside graph

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95920
* __->__ #95893
* #95890

Inductor implementations of collectives/wait must match
eager impls in _functional_collectives in terms of interacting
with _register_tensor_work API.  If they do, then splitting
a collective-wait pair so one half is in a compiled graph should
work fine.

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: test/distributed/test_inductor_collectives.py [hunk 1]
```diff
@@ -136,7 +212,8 @@ def func(inp, *, tag, ranks, group_size):
             .check("buf0 = empty_strided") \
             .check("buf0.copy_(arg0_1)") \
             .check("buf0_work = dist.all_reduce(buf0") \
-            .check("buf0_work.wait()") \
+            .check("_register_tensor_work(buf0, buf0_work)") \
```
Comment (kumpera) [inline test/distributed/test_inductor_collectives.py:215]: We should later add comments to _functional_collectives mentioning that those methods are used by inductor

**Review by kumpera — APPROVED** [review]
Looks good, the only concern is that we need to eventually establish an API that inductor can depend on instead of using internals as is today.

**Comment by wconstab** [comment]
@pytorchhbot merge

**Pull request closed without merging**